agent/xds/delta.go

```diff
@@ -0,0 +1,728 @@
+		s.CheckFetcher,
+		s.CfgFetcher,
+		true,
+	)
+
+	// need to run a small state machine to get through initial authentication.
+	var state = stateDeltaInit
+
+	// Configure handlers for each type of request we currently care about.
+	handlers := map[string]*xDSDeltaType{
+		ListenerType: newDeltaType(g, stream, ListenerType, func(kind structs.ServiceKind) bool {
+			return cfgSnap.Kind == structs.ServiceKindIngressGateway
+		}),
+		RouteType: newDeltaType(g, stream, RouteType, func(kind structs.ServiceKind) bool {
+			return cfgSnap.Kind == structs.ServiceKindIngressGateway
+		}),
+		ClusterType: newDeltaType(g, stream, ClusterType, func(kind structs.ServiceKind) bool {
+			// Mesh, Ingress, and Terminating gateways are allowed to inform CDS of
+			// no clusters.
+			return cfgSnap.Kind == structs.ServiceKindMeshGateway ||
+				cfgSnap.Kind == structs.ServiceKindTerminatingGateway ||
+				cfgSnap.Kind == structs.ServiceKindIngressGateway
+		}),
+		EndpointType: newDeltaType(g, stream, EndpointType, nil),
+	}
+
+	var deltaRetryFrequency = s.DeltaRetryFrequency
+	if deltaRetryFrequency == 0 {
+		deltaRetryFrequency = DefaultDeltaRetryFrequency
+	}
+
+	var retryTimer <-chan time.Time
+	extendRetryTimer := func() {
+		g.Logger.Trace("retrying response", "after", deltaRetryFrequency)
+		retryTimer = time.After(deltaRetryFrequency)
+	}
+
+	var authCheckFrequency = s.AuthCheckFrequency
+	if authCheckFrequency == 0 {
+		authCheckFrequency = DefaultAuthCheckFrequency
+	}
+
+	var authTimer <-chan time.Time
+	extendAuthTimer := func() {
+		authTimer = time.After(authCheckFrequency)
+	}
+
+	checkStreamACLs := func(cfgSnap *proxycfg.ConfigSnapshot) error {
+		return s.checkStreamACLs(stream.Context(), cfgSnap)
+	}
+
+	for {
+		select {
+		case <-authTimer:
+			// It's been too long since a Discovery{Request,Response} so recheck ACLs.
+			if err := checkStreamACLs(cfgSnap); err != nil {
+				return err
+			}
+			extendAuthTimer()
+
+		case req, ok := <-reqCh:
+			if !ok {
+				// reqCh is closed when stream.Recv errors which is how we detect client
+				// going away. AFAICT the stream.Context() is only canceled once the
+				// RPC method returns which it can't until we return from this one so
+				// there's no point in blocking on that.
+				return nil
+			}
+
+			g.logTraceRequest("Incremental xDS v3", req)
+
+			if req.TypeUrl == "" {
+				return status.Errorf(codes.InvalidArgument, "type URL is required for ADS")
+			}
+
+			handler, ok := handlers[req.TypeUrl]
+			if !ok {
+				return nil // not a type we care about
+			}
+			if handler.Recv(req) {
+				g.Logger.Trace("subscribing to type", "typeUrl", req.TypeUrl)
+			}
+
+			if node == nil && req.Node != nil {
+				node = req.Node
+				var err error
+				g.ProxyFeatures, err = determineSupportedProxyFeatures(req.Node)
+				if err != nil {
+					return status.Errorf(codes.InvalidArgument, err.Error())
+				}
+			}
+
+		case cfgSnap = <-stateCh:
+			newRes, err := g.allResourcesFromSnapshot(cfgSnap)
```

Comment:
This is the first big divergence from SoTW: instead of delaying xDS construction until send-time, we pre-generate it when the snapshot changes so we can compute versions to use when doing the differential sync.